Feat: 스터디룸 생성 시 단계별 시간 설정 기능 추가

kkkkkSE opened the pull request:
- 스터디룸 생성 시 단계별 시간 설정 기능 추가
- 스터디룸 생성 시 단계별 시간 설정 기능 테스트 코드 작성
  - MUI의 Select 컴포넌트 특성 상 select, option 태그를 사용하지 않고 ul, li 태그를 사용하고 있어 차후 다른 UI 라이브러리(시멘틱 태그) 적용 후 일부 테스트 코드 추가 예정

kkkkkSE committed:
Feat: 스터디룸 생성 시 단계별 시간 설정 기능 추가

kkkkkSE committed:
Test: 스터디룸 생성 시 시간 설정 기능 관련 테스트 작성

jungjinbeom commented on src/pages/study-rooms/components/create-study-room-selects.tsx:
```diff
@@ -0,0 +1,116 @@
+          <MenuItem key={timeValue} value={timeValue}>
+            {timeValue}분
+          </MenuItem>
+        ))}
```
스터디시간, 계획시간, 회고시간, 휴식 시간 별로 시간값을 가지고 같은 UI를 만들어주는거 같은데 따로 컴포넌트로 빼는건 어떨까요??

Seung-wan commented on src/components/common/input/select-box.tsx:
```diff
@@ -0,0 +1,17 @@
+import { PropsWithChildren } from 'react';
+import { FormControl, InputLabel, Select, SelectProps } from '@mui/material';
+
+interface SelectBoxProps extends Omit<SelectProps, 'labelId'> {
+  labelId: string;
+}
```
labelId가 mui 내부에서는 어떤 타입으로 잡혀있나요?

kkkkkSE replied:
타입은 위와 동일하게 string이 맞는데 mui 내부에선 옵셔널하게 받고 있어서 필수로 입력하도록 변경했습니다~!

Seung-wan commented on src/components/common/input/select-box.tsx:
```diff
@@ -0,0 +1,17 @@
+  labelId: string;
+}
+
+const SelectBox = ({ children, ...props }: PropsWithChildren<SelectBoxProps>) => {
+  return (
+    <FormControl fullWidth>
+      <InputLabel id={props.labelId}>{props.label}</InputLabel>
+      <Select {...props}>{children}</Select>
```
이렇게 표현해주는건 어떨까요?
```suggestion
const SelectBox = ({ children, labelId, label, ...restProps }: PropsWithChildren<SelectBoxProps>) => {
  return (
    <FormControl fullWidth>
      <InputLabel id={labelId}>{label}</InputLabel>
      <Select {...restProps}>{children}</Select>
```

Seung-wan commented on src/components/common/input/select-box.tsx:
저는 select라는 컴포넌트는 input과 동등한 계층으로 존재하는게 적절하다고 생각하는데 (input의 하위가 아닌) 어떻게 생각하시나요?

kkkkkSE replied:
사용자의 입력을 받는다는 의미에서 input 폴더 내에 위치하도록 해뒀는데, 생각해보니 태그 자체도 input과 select는 별개여서 의견주신 부분이 더 맞다는 생각이 드네요 🤔

Seung-wan commented on src/pages/study-rooms/components/create-study-room-modal.tsx:
```diff
@@ -1,43 +1,23 @@
-    } catch (e) {
-      console.error(e);
-    }
+  const handleClickCreateStudyRoom = async () => {
+    createStudyRoom({ name, intro, timeSet });
```
async를 붙여줄 필요가 있나요?

kkkkkSE replied:
앗 그렇네요 수정하겠습니다 ㅎ

Seung-wan commented on src/pages/study-rooms/utils/generate-time-values.ts:
```diff
@@ -0,0 +1,9 @@
+import { STUDY_ROOM_STEP_TIME_UNIT } from '@/constant/study-room';
+
+const generateTimeValues = (min: number, max: number) => {
+  const length = Math.floor((max - min) / STUDY_ROOM_STEP_TIME_UNIT) + 1;
+
+  return Array.from({ length }, (_, index) => min + index * STUDY_ROOM_STEP_TIME_UNIT);
+};
+
+export default generateTimeValues;
```
`src/pages/study-rooms/components/create-study-room-selects.tsx` 파일 내부에 작성해줘도 괜찮지 않을까요?

Seung-wan commented on src/pages/study-rooms/components/create-study-room-selects.tsx:
```diff
@@ -0,0 +1,116 @@
+  onChangeSelect: (event: SelectChangeEvent<unknown>) => void;
+}
+
+const CreateStudyRoomSelects = ({ timeSet, onChangeSelect }: CreateStudyRoomSelectsProps) => {
```
Selects 보다는 SelectFields 같은 느낌으로 표현해주면 어떨까요?
ui를 묘사해주면 더 좋을 것 같아요~

Seung-wan commented on src/pages/study-rooms/components/create-study-room-selects.tsx:
```diff
@@ -0,0 +1,116 @@
+
+const CreateStudyRoomSelects = ({ timeSet, onChangeSelect }: CreateStudyRoomSelectsProps) => {
+  // step별 시간 선택 select 요소로 들어갈 value 값을 배열로 저장한 변수
+  const stepTimeValues = useMemo(
+    () =>
+      (() => {
+        const planning = generateTimeValues(
+          STUDY_ROOM_STEP_PROGRESS_TIME.PLANNING.MIN,
+          STUDY_ROOM_STEP_PROGRESS_TIME.PLANNING.MAX,
+        );
+
+        const studying = generateTimeValues(
+          STUDY_ROOM_STEP_PROGRESS_TIME.STUDYING.MIN,
+          STUDY_ROOM_STEP_PROGRESS_TIME.STUDYING.MAX,
+        );
+
+        const retrospect = generateTimeValues(
+          STUDY_ROOM_STEP_PROGRESS_TIME.RETROSPECT.MIN,
+          STUDY_ROOM_STEP_PROGRESS_TIME.RETROSPECT.MAX,
+        );
+
+        const resting = generateTimeValues(
+          STUDY_ROOM_STEP_PROGRESS_TIME.RESTING.MIN,
+          STUDY_ROOM_STEP_PROGRESS_TIME.RESTING.MAX,
+        );
+
+        return { planning, studying, retrospect, resting };
+      })(),
+    [],
+  );
```
컴포넌트 외부에서 만들어줘도 괜찮지 않을까요?

kkkkkSE replied:
만약 그렇게 하면 동작에 차이가 있나용 ? ?

Seung-wan replied:
지금은 useMemo의 의존성이 빈배열이다보니 한번만 생성해놓고 이후에는 계속 사용할 수 있을 것 같은데요~
아예 외부에서 해당 값을 관리해주면 useMemo를 실행하는 비용조차도 줄일 수 있을 것 같아서 드린 의견입니다!

kkkkkSE committed:
Refactor: 스터디룸 생성에 사용하는 Select 컴포넌트 형태 변경
- 기존 : Select 공통 컴포넌트를 이용, Option은 children으로 받음
- 변경 : Select 공통 컴포넌트와 비슷한 형태인 StepTimeSelect 컴포넌트 추가, Option Values를 Props로 받음

kkkkkSE committed:
Style: 코드리뷰를 반영한 단순 코드 스타일, 네이밍 변경

Seung-wan commented on src/pages/study-rooms/components/create-study-room-modal.tsx:
```diff
@@ -1,43 +1,23 @@
 import useModal from '@/hooks/useModal';
+import useCreateStudyRoomForm from '../hooks/useCreateStudyRoomForm';
 import Modal from '@/components/common/modal/modal';
+import CreateStudyRoomSelects from './create-study-room-select-fields';
```
컴포넌트 이름을 그대로 import해오면 좋을 것 같네요~
```suggestion
import CreateStudyRoomSelectFields from './create-study-room-select-fields';
```

Seung-wan replied:
근데 해당 컴포넌트 내에서 표현되는 UI를 보니까 지금 작성된 이름보다 조금 더 명확하게 작성해볼만 한 것 같아요~ StepTime을 선택하는 필드들을 모아놓은 것 같아서요!
실제로 select 자체가 스터디룸을 생성하는 역할(지금은 생성하는데 필요한 부품 정도?)을 하는건 아니기도 하구요~

Seung-wan commented on src/pages/study-rooms/components/create-study-room-modal.tsx:
```diff
@@ -46,26 +26,27 @@ const CreateStudyRoomModal = () => {
+        name="intro"
+        variant="outlined"
+        multiline
+        rows={3}
```
요것도 상수화해봐도 괜찮겠네요~

Seung-wan commented on src/pages/study-rooms/components/create-study-room-select-fields.tsx:
```diff
@@ -0,0 +1,99 @@
+}
+
+// step별 시간 선택 select 요소로 들어갈 value 값을 배열로 저장한 변수
+const stepTimeValues = (() => {
+  const generateTimeValues = (min: number, max: number) => {
+    const length = Math.floor((max - min) / STUDY_ROOM_STEP_TIME_UNIT) + 1;
+
+    return Array.from({ length }, (_, index) => min + index * STUDY_ROOM_STEP_TIME_UNIT);
+  };
+
+  const planning = generateTimeValues(
+    STUDY_ROOM_STEP_PROGRESS_TIME.PLANNING.MIN,
+    STUDY_ROOM_STEP_PROGRESS_TIME.PLANNING.MAX,
+  );
+
+  const studying = generateTimeValues(
+    STUDY_ROOM_STEP_PROGRESS_TIME.STUDYING.MIN,
+    STUDY_ROOM_STEP_PROGRESS_TIME.STUDYING.MAX,
+  );
+
+  const retrospect = generateTimeValues(
+    STUDY_ROOM_STEP_PROGRESS_TIME.RETROSPECT.MIN,
+    STUDY_ROOM_STEP_PROGRESS_TIME.RETROSPECT.MAX,
+  );
+
+  const resting = generateTimeValues(
+    STUDY_ROOM_STEP_PROGRESS_TIME.RESTING.MIN,
+    STUDY_ROOM_STEP_PROGRESS_TIME.RESTING.MAX,
+  );
+
+  return { planning, studying, retrospect, resting };
+})();
```
이해하는데 큰 어려움이 있는 건 아니긴 한데요!
IIFE 내부에 함수가 있는 구조라서 어떤 값이 만들어지는지 명확하게 보이는 건 아니라고 생각해서 아래와 같이 표현해주면 어떨까 싶네요~

```suggestion
const generateTimeValues = (min: number, max: number) => {
  const length = Math.floor((max - min) / STUDY_ROOM_STEP_TIME_UNIT) + 1;

  return Array.from(
    { length },
    (_, index) => min + index * STUDY_ROOM_STEP_TIME_UNIT,
  );
};

const STEP_TIME_VALUES = {
  PLANNING: generateTimeValues(
    STUDY_ROOM_STEP_PROGRESS_TIME.PLANNING.MIN,
    STUDY_ROOM_STEP_PROGRESS_TIME.PLANNING.MAX,
  ),
  STUDYING: generateTimeValues(
    STUDY_ROOM_STEP_PROGRESS_TIME.STUDYING.MIN,
    STUDY_ROOM_STEP_PROGRESS_TIME.STUDYING.MAX,
  ),
  RETROSPECT: generateTimeValues(
    STUDY_ROOM_STEP_PROGRESS_TIME.RETROSPECT.MIN,
    STUDY_ROOM_STEP_PROGRESS_TIME.RETROSPECT.MAX,
  ),
  RESTING: generateTimeValues(
    STUDY_ROOM_STEP_PROGRESS_TIME.RESTING.MIN,
    STUDY_ROOM_STEP_PROGRESS_TIME.RESTING.MAX,
  ),
};

```

kkkkkSE replied:
generateTimeValues가 오로지 step time 배열을 만드는데만 사용돼서 함수 안에 넣은거였는데, 분리해주는게 보기엔 더 깔끔하긴 하네요 🤔

Seung-wan commented on src/pages/study-rooms/components/create-study-room-select-fields.tsx:
```diff
@@ -0,0 +1,99 @@
+  return (
+    <SelectsWrapper>
+      <StepTimeSelect
+        labelId="planning-time-select"
```
labelId는 케밥케이스로 표현해주신 이유가 있으실까요? name은 카멜케이스여서요~

kkkkkSE replied:
저는 보통 id나 class 등을 표현할때 케밥 케이스를 사용하는데요, name을 카멜 케이스로 작성한 이유는 select 핸들러 함수에서 name 속성을 이용하여 (useCreateStudyRoomForm 훅 안에 있는) timeSet state를 변경시키고 있기 때문입니다 🙂

Seung-wan commented on src/pages/study-rooms/components/step-time-select.tsx:
```diff
@@ -0,0 +1,29 @@
+
+interface StepTimeSelectProps extends Omit<SelectProps, 'labelId'> {
+  labelId: string;
+  optionValues: number[];
```
컴포넌트 이름이 StepTime을 선택하는 Select라고 표현이 되어 있어서 optionValues보다는 stepTime이라는 데이터 이름을 드러내줘도 괜찮을 것 같네요~

Seung-wan commented on src/pages/study-rooms/hooks/useCreateStudyRoomForm.tsx:
```diff
@@ -0,0 +1,72 @@
+const useCreateStudyRoomForm = () => {
+  const navigate = useNavigate();
+
+  const { onClose } = useModal(MODAL_KEYS.CREATE_STUDY_ROOM);
```
모달에 대한 맥락없이 여기서 뜬금없이 모달을 닫는 훅이 사용되고 있는게 조금 어색한 것 같아요.
보통은 모달 컴포넌트에서 어떤 버튼을 클릭했을 때 props로 받은 onClose 함수를 실행시켜서 모달을 닫는식으로 많이 구현해서요.

Seung-wan commented on src/pages/study-rooms/hooks/useCreateStudyRoomForm.tsx:
```diff
@@ -0,0 +1,72 @@
+import { TimeSet } from '@/types/study-room.types';
+import useModal from '@/hooks/useModal';
+
+const useCreateStudyRoomForm = () => {
```
현재 훅이 스터디룸 Form을 관리하는 역할이랑, 스터디룸을 생성하는 역할 두가지를 하고 있는데 요건 같이둬도 괜찮을지 고민해보면 좋을 것 같습니다~

kkkkkSE replied:
Form을 관리하는 역할 중 하나가 Submit과 같은 동작도 포함돼있다고 생각해서 스터디룸 생성 요청 함수를 훅 안에다가 넣어뒀는데, 이것 때문에 위에서 말씀하신 onClose, navigate가 같이 훅 내부로 들어와서 뭔가 부자연스럽다는 느낌에 고민을 많이 했었어요 ㅠ 스터디룸 생성을 분리해도 괜찮다는 의견이시라면 그 방향이 훨씬 좋을 것 같긴 하네요 🤔

kkkkkSE committed:
Refactor: 스터디룸 생성 함수 위치 변경

kkkkkSE committed:
Style: 코드리뷰를 반영한 단순 코드 스타일, 네이밍 변경
